src/connector/src/sink/big_query.rs

```diff
@@ -582,80 +565,64 @@ impl BigQuerySinkWriter {
         }
         Ok(serialized_rows)
     }
-
-    async fn write_rows(&mut self) -> Result<()> {
-        if self.write_rows.is_empty() {
-            return Ok(());
-        }
-        let mut errs = Vec::with_capacity(self.config.common.retry_times);
-        for _ in 0..self.config.common.retry_times {
-            match self
-                .client
-                .append_rows(self.write_rows.clone(), self.write_stream.clone())
-                .await
-            {
-                Ok(_) => {
-                    self.write_rows_count = 0;
-                    self.write_rows.clear();
-                    return Ok(());
-                }
-                Err(e) => errs.push(e),
-            }
-        }
-        Err(SinkError::BigQuery(anyhow::anyhow!(
-            "Insert error {:?}",
-            errs
-        )))
-    }
 }
 
-#[async_trait]
-impl SinkWriter for BigQuerySinkWriter {
-    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+pub type BigQuerySinkDeliveryFuture = impl TryFuture<Ok = (), Error = SinkError> + Unpin + 'static;
+
+impl AsyncTruncateSinkWriter for BigQuerySinkWriter {
+    type DeliveryFuture = BigQuerySinkDeliveryFuture;
+
+    async fn write_chunk<'a>(
+        &'a mut self,
+        chunk: StreamChunk,
+        mut add_future: DeliveryFutureManagerAddFuture<'a, Self::DeliveryFuture>,
+    ) -> Result<()> {
         let serialized_rows = if self.is_append_only {
             self.append_only(chunk)?
         } else {
             self.upsert(chunk)?
         };
-        if !serialized_rows.is_empty() {
-            self.write_rows_count += serialized_rows.len();
-            let rows = AppendRowsRequestRows::ProtoRows(ProtoData {
-                writer_schema: Some(self.writer_pb_schema.clone()),
-                rows: Some(ProtoRows { serialized_rows }),
-            });
-            self.write_rows.push(rows);
-
-            if self.write_rows_count >= self.config.common.max_batch_rows {
-                self.write_rows().await?;
-            }
-        }
-        Ok(())
-    }
-
-    async fn begin_epoch(&mut self, _epoch: u64) -> Result<()> {
-        Ok(())
-    }
-
-    async fn abort(&mut self) -> Result<()> {
-        Ok(())
-    }
-
-    async fn barrier(&mut self, is_checkpoint: bool) -> Result<()> {
-        if is_checkpoint {
-            self.write_rows().await?;
+        if serialized_rows.is_empty() {
+            return Ok(());
         }
-        Ok(())
-    }
-
-    async fn update_vnode_bitmap(&mut self, _vnode_bitmap: Arc<Bitmap>) -> Result<()> {
+        let rows = AppendRowsRequestRows::ProtoRows(ProtoData {
+            writer_schema: Some(self.writer_pb_schema.clone()),
+            rows: Some(ProtoRows { serialized_rows }),
+        });
+        let (expect_offset, mut rx) = self.client.get_subscribe();
+        let future = Box::pin(async move {
+            loop {
+                match rx.recv().await {
```

Comment:
I think instead you can create a oneshot channel for each AppendRowsRequest  so that you won't need this shared broadcast channel. When the spawned worker knows that a request is handled, it notifies the rx with the oneshot tx, and here the future can be simply the rx.